src/algorithms/calorimetry/ImagingTopoCluster.cc

```diff
@@ -191,20 +246,65 @@ bool ImagingTopoCluster::is_neighbour(const edm4eic::CalorimeterHit& h1,
+                       edm4hep::utils::eta(h2.getPosition())) <= sameLayerDistEtaPhi[0]) &&
+             (std::abs(edm4hep::utils::angleAzimuthal(h1.getPosition()) -
+                       edm4hep::utils::angleAzimuthal(h2.getPosition())) <= sameLayerDistEtaPhi[1]);
+
+    case ImagingTopoClusterConfig::ELayerMode::phiz: {
+      // Layer mode 'phiz' uses the average phi of the hits to define a rotated direction. The coordinate is a distance, not an angle.
+      auto phi  = 0.5 * (edm4hep::utils::angleAzimuthal(h1.getPosition()) +
+                        edm4hep::utils::angleAzimuthal(h2.getPosition()));
+      auto h1_t = (h1.getPosition().x * sin(phi)) - (h1.getPosition().y * cos(phi));
+      auto h2_t = (h2.getPosition().x * sin(phi)) - (h2.getPosition().y * cos(phi));
+      auto h1_z = h1.getPosition().z;
+      auto h2_z = h2.getPosition().z;
+
+      return (std::abs(h1_t - h2_t) <= sameLayerDistPhiZ[0]) &&
+             (std::abs(h1_z - h2_z) <= sameLayerDistPhiZ[1]);
+    }
+
+    default:
+      error("Unknown layer mode for same-layer clustering.");
+      return false;
+    }
   } else if (ldiff <= m_cfg.neighbourLayersRange) {
-    switch (m_cfg.layerMode) {
+    switch (m_cfg.diffLayerMode) {
     case eicrecon::ImagingTopoClusterConfig::ELayerMode::etaphi:
       return (std::abs(edm4hep::utils::eta(h1.getPosition()) -
-                       edm4hep::utils::eta(h2.getPosition())) <= layerDistEtaPhi[0]) &&
+                       edm4hep::utils::eta(h2.getPosition())) <= diffLayerDistEtaPhi[0]) &&
              (std::abs(edm4hep::utils::angleAzimuthal(h1.getPosition()) -
-                       edm4hep::utils::angleAzimuthal(h2.getPosition())) <= layerDistEtaPhi[1]);
+                       edm4hep::utils::angleAzimuthal(h2.getPosition())) <= diffLayerDistEtaPhi[1]);
+
     case eicrecon::ImagingTopoClusterConfig::ELayerMode::xy:
-      return (std::abs(h1.getPosition().x - h2.getPosition().x) <= layerDistXY[0]) &&
-             (std::abs(h1.getPosition().y - h2.getPosition().y) <= layerDistXY[1]);
+      // Here, the xy layer mode is based on global XY positions rather than local XY positions, and thus it only works for endcap detectors.
+      return (std::abs(h1.getPosition().x - h2.getPosition().x) <= diffLayerDistXY[0]) &&
+             (std::abs(h1.getPosition().y - h2.getPosition().y) <= diffLayerDistXY[1]);
+
+    case eicrecon::ImagingTopoClusterConfig::ELayerMode::phiz: {
+      auto phi  = 0.5 * (edm4hep::utils::angleAzimuthal(h1.getPosition()) +
+                        edm4hep::utils::angleAzimuthal(h2.getPosition()));
+      auto h1_t = (h1.getPosition().x * sin(phi)) - (h1.getPosition().y * cos(phi));
+      auto h2_t = (h2.getPosition().x * sin(phi)) - (h2.getPosition().y * cos(phi));
+      auto h1_z = h1.getPosition().z;
+      auto h2_z = h2.getPosition().z;
+
+      return (std::abs(h1_t - h2_t) <= diffLayerDistPhiZ[0]) &&
+             (std::abs(h1_z - h2_z) <= diffLayerDistPhiZ[1]);
```

Comment:
I would rename `PhiZ` to `TZ` (and `phiz` to `tz`). You'll be able to remove some explanatory comments too.